File: servers/kt-ktor/src/main/kotlin/com/bench/ktor/routes/WebRoutes.kt
package com.bench.ktor.routes

import com.bench.ktor.ApiException
import com.bench.ktor.appJson
import com.bench.shared.Consts
import com.bench.shared.model.OrderValidation
import com.bench.shared.model.ValidateRequest
import com.bench.shared.web.Compute
import com.bench.shared.web.Jwt
import io.ktor.server.freemarker.FreeMarkerContent
import io.ktor.server.request.receiveText
import io.ktor.server.response.respond
import io.ktor.server.routing.Route
import io.ktor.server.routing.get
import io.ktor.server.routing.post
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

@Serializable
private data class TokenResponse(
    val token: String,
)

// Contract canon for GET /html (rendered into index.ftl): greeting name, fruit
// list, and the labeled total the htmlContains assertion checks for.
private const val HTML_NAME = "Alice"
private const val HTML_TOTAL = 42
private val HTML_FRUITS = listOf("apple", "banana", "cherry")

/** The web suite: `GET /html`, `GET /jwt/sign`, `GET /jwt/verify`, `POST /validate`, `GET /compute`. */
fun Route.webRoutes(jwtSecret: String) {
    get("/html") {
        call.respond(
            FreeMarkerContent(
                "index.ftl",
                mapOf("name" to HTML_NAME, "fruits" to HTML_FRUITS, "total" to HTML_TOTAL),
            ),
        )
    }

    get("/jwt/sign") {
        call.respond(TokenResponse(Jwt.sign(jwtSecret)))
    }

    get("/jwt/verify") {
        val header = call.request.headers["Authorization"]
        if (header == null || !header.startsWith("Bearer ")) {
            throw ApiException.Unauthorized(Consts.ERR_INVALID_TOKEN)
        }
        val claims =
            Jwt.verify(jwtSecret, header.removePrefix("Bearer "))
                ?: throw ApiException.Unauthorized(Consts.ERR_INVALID_TOKEN)
        call.respond(claims)
    }

    post("/validate") {
        // Decode failures are malformed JSON / type mismatches ONLY (every schema
        // field has a zero-value default, Go canon) — so they map to the invalid-JSON
        // body error, mirroring Go's WriteBodyError; rule breaches are the
        // validator's job below.
        val request =
            try {
                appJson.decodeFromString<ValidateRequest>(call.receiveText())
            } catch (e: SerializationException) {
                throw ApiException.BadRequest(Consts.ERR_INVALID_JSON, e.message ?: "invalid payload", e)
            }
        val errors = OrderValidation.validate(request)
        if (errors.isEmpty()) {
            call.respond(buildJsonObject { put("valid", true) })
        } else {
            throw ApiException.BadRequest(Consts.ERR_VALIDATION_FAILED, errors.joinToString("; "))
        }
    }

    get("/compute") {
        val n = call.request.queryParameters["n"]?.toLongOrNull()
        if (n == null || n < 1) throw ApiException.BadRequest(Consts.ERR_INVALID_N, "n must be an integer >= 1")
        val result = withContext(Dispatchers.Default) { Compute.sha256Chain(Compute.clampRounds(n)) }
        call.respond(buildJsonObject { put("result", result) })
    }
}

File: shared/typescript-hono/src/web-routes.ts
// /web routes: the five web-suite endpoints (/html, /jwt/sign, /jwt/verify,
// /validate, /compute), shared by the three Hono runtime entries (ts-honojs on
// Node, ts-bun-honojs on Bun, ts-deno-honojs on Deno). All the logic — the
// validate rules, the SHA-256 chain, the jose JWT helpers — lives in @bench/shared;
// these handlers are the Hono-idiomatic glue only (web-standard Request/Response,
// so identical across runtimes).

import {
  env,
  INVALID_JSON_BODY,
  INVALID_N,
  INVALID_TOKEN,
  makeError,
  parseComputeRounds,
  sha256Chain,
  signToken,
  VALIDATION_FAILED,
  validateWebPayload,
  verifyToken
} from "@bench/shared";
import { Hono } from "hono";

// The /html canon: greeting + fruit list + labeled total, rendered as a
// server-side template literal. The contract matches each value with
// htmlContains (whitespace-tolerant), not a byte-exact body.
const FRUITS = ["apple", "banana", "cherry"];

function renderHtml(): string {
  const items = FRUITS.map((fruit) => `    <li>${fruit}</li>`).join("\n");
  return `<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"><title>Benchmark</title></head>
<body>
  <h1>Hello, Alice</h1>
  <ul>
${items}
  </ul>
  <p>Total: 42</p>
</body>
</html>`;
}

export function createWebRoutes(): Hono {
  const webRoutes = new Hono();

  webRoutes.get("/html", (c) => c.html(renderHtml()));

  webRoutes.get("/jwt/sign", async (c) => {
    const token = await signToken(env.JWT_SECRET);
    return c.json({ token });
  });

  webRoutes.get("/jwt/verify", async (c) => {
    const auth = c.req.header("Authorization") ?? "";
    const token = auth.startsWith("Bearer ") ? auth.slice(7).trim() : "";
    if (!token) return c.json(makeError(INVALID_TOKEN, "missing bearer token"), 401);
    try {
      const claims = await verifyToken(env.JWT_SECRET, token);
      return c.json(claims);
    } catch (err) {
      return c.json(makeError(INVALID_TOKEN, err), 401);
    }
  });

  webRoutes.post("/validate", async (c) => {
    let body: unknown;
    try {
      body = await c.req.json();
    } catch (err) {
      return c.json(makeError(INVALID_JSON_BODY, err), 400);
    }

    const result = validateWebPayload(body);
    if (!result.ok) return c.json(makeError(VALIDATION_FAILED, result.details), 400);
    return c.json({ valid: true });
  });

  webRoutes.get("/compute", (c) => {
    const rounds = parseComputeRounds(c.req.query("n"));
    if (rounds === null) return c.json(makeError(INVALID_N, "n must be an integer >= 1"), 400);
    return c.json({ result: sha256Chain(rounds) });
  });

  return webRoutes;
}
